### PR TITLE
Port to GNU MPFR/MPC Library

### PR DESCRIPTION
**Pull request by mbkma**
As part of a hobby project I ported mate-calc to the GNU MPFR/MPC Library. Advantages over mate-calcs internal library include: mpfr/mpc supports **much more functions**, is very **well documented** and features **more efficient algorithms**. Generally speaking it is now possible to compute very small and very large operands with better precision in less time.
For example it is now possible to calculate:
`ln(10^-10000000)`, `asin(i)`, `factorize 1234567891011121314`, `1234^1234567`, `0.5!`

It also **fixes some bugs** in mate-calc master I noticed during porting:
`ln(0.5)` does not converge
`321^3124` crashes mate-calc in fixed mode

I tried to keep as much as possible from the original program structure. The only "new" things include the possibility to calculate factorials of positive real numbers and an integer based factorization function to improve factorization performance for integers < 2^64-1 based on gnome-calcs implementation. Mostly the added code initializes and clears MPNumbers, because this is necessary now, see MPCs documentation for details. As of regressions I noticed the spinner icon missing while computing large powers.
I would have informed you before about this PR, as it is suggested on the mate-wiki, but I couldn't find an active IRC or mailing-list. Still, I am looking forward to your feedback.

For further information please visit:
https://www.mpfr.org/ (mpfr library)
http://www.multiprecision.org/mpc (mpc library)
https://en.wikipedia.org/wiki/Gamma_function (In mathematics, the gamma function is one commonly used extension of the factorial function to complex numbers)
https://www.wolframalpha.com/ (wolframalpha, for testing purposes)

### CODE REVIEW

**Comment by mbkma** [comment]
@rbuj I don't know exactly what you are referring to, but for the test-cases the external mpc_to_string method is used. It has nothing to do with the actual output of mate-calc. Try computing those numbers in mate-calc, you will see that there is no loss of precision whatsoever.
If it is important I could restore the output of test cases to match the current behaviour in master.

**Comment by sc0w** [comment]
fedora still fails, @mbkma please add the packages in alphabetical order

**Comment by mbkma** [comment]
Why does fedora ship such an outdated version of mpfr?
From travis:
` mpfr-devel                   x86_64  3.1.6-5.fc31               fedora    75 k`
Newest version is 4.0.2.

**Comment by raveit65** [comment]
fedora 32 will have mpfr-4.0.2 in May
https://koji.fedoraproject.org/koji/packageinfo?packageID=4830

**Comment by raveit65** [comment]
@mbkma @sc0w 
I agree with disable temporarily fedora builds at travis for master branch, until fedora 32 is released.
Just comment out that line `DISTRO="fedora:latest"` at https://github.com/mate-desktop/mate-calc/blob/master/.travis.yml#L58
Is that helpful?

**Comment by raveit65** [comment]
@mbkma 
Or better change `DISTRO="fedora:latest"` to `DISTRO="fedora:32"`.
Fedora 32 is already in beta freeze and will ship MATE 1.24.

**Comment by raveit65** [comment]
Hmm,
looks like shit can always be happened before fedora 32 beta release :P
```
Running scriptlet: systemd-245~rc1-4.fc32.x86_64                        49/94 

  Upgrading        : systemd-245~rc1-4.fc32.x86_64                        49/94 

  Running scriptlet: systemd-245~rc1-4.fc32.x86_64                        49/94 

The command "./docker-build --name ${DISTRO} --config .build.yml --install" failed and exited with 1 during .
```
No idea, why installing systemd-245~rc1-4.fc32.x86_64 failed in docker cloud.
@mbkma 
Simply comment out this line and disable fedora build at travis temporarily.
All this pain will go away by magic when f32 is stable in a few weeks.
Only releasing 1.25 dev releases is impossible without fedora.
Not real problem for some weeks.

**Comment by sc0w** [comment]
@rbuj @mbkma it is possible to check mpfr version in configure.ac? and we can check if 4.0.2 is installed, I think it can be helpful for the other distros.

**Comment by mbkma** [comment]
@sc0w do you mean something like this?

**Comment by sc0w** [comment]
yes, thanks

I think the commits can be reorder a bit:

https://github.com/mate-desktop/mate-calc/pull/138/commits/50612a55555051a600bffa60da77855e5da1a690 and https://github.com/mate-desktop/mate-calc/pull/138/commits/4d12a88e7b222e42a6debb8221e46ecee7a19d5f can be squashed into first commit

https://github.com/mate-desktop/mate-calc/pull/138/commits/72d62d6e67b1ee4cff169097f62eecdaf6060b65 and https://github.com/mate-desktop/mate-calc/pull/138/commits/f705201ea1a87b73106febacd14665b8cd33bae6 can be squashed into first commit and "travis" or "ci" can be mentioned in commit message

in commit https://github.com/mate-desktop/mate-calc/pull/138/commits/66884c66d421cc393540d68040b42125dc5e0acc "travis" or "ci" can be mentioned in commit message

**Comment by raveit65** [comment]
Only for the record.
MPFR_REQUIRED=4.0.2 means that a build will fail if the require isn`t satisfied.
Does that work for stable debian (mostly oldest pkdb-base), slackware/gentoo or BSD distros?

**Comment by mbkma** [comment]
On Debian stable it works, of slackware/gentoo or BSD distros I have no idea.

**Comment by raveit65** [comment]
> On Debian stable it works, of slackware/gentoo or BSD distros I have no idea.

Good, at least this work is a new feature for master development, so it might be OK for all distros to target it for 1.26 next year.

File: src/unittest.c [hunk 1]
```diff
@@ -110,6 +110,7 @@ static void test(char* expression, char* expected, int expected_error)
 			fail("'%s' -> error %s, expected error %s", expression, error_code_to_string(error), error_code_to_string(expected_error));
 		}
 	}
+    mp_clear(result);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

File: src/unittest.c [hunk 2]
```diff
@@ -594,54 +595,35 @@ static void test_equations()
...
-
-		if (j == MP_SIZE)
-		{
-			printf(",...");
-			break;
-		}
-
-		printf(",%d", x->fraction[i]);
-	}
+    mpc_out_str(stdout, 10, 5,  x->num, MPC_RNDNN);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

File: src/unittest.c [hunk 3]
```diff
@@ -594,54 +595,35 @@ static void test_equations()
...
 {
-	MPNumber t;
+	MPNumber t = mp_new();
 
 	mp_set_from_string(number, 10, &t);
 
 	printf("MPNumber(%s) -> {", number);
 	print_number(&t);
 	printf("}\n");
+    mp_clear(&t);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

File: src/unittest.c [hunk 4]
```diff
@@ -594,54 +595,35 @@ static void test_equations()
...
 {
-	MPNumber t;
+	MPNumber t = mp_new();
 
 	mp_set_from_integer(number, &t);
 
 	printf("MPNumber(%d) -> {", number);
 	print_number(&t);
 	printf("}\n");
+    mp_clear(&t);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

File: src/unittest.c [hunk 5]
```diff
@@ -687,11 +669,12 @@ static void try(const char* string, bool result, bool expected)
...
-
-	mp_set_from_integer(0, &zero);
-	mp_set_from_integer(1, &one);
-	mp_set_from_integer(-1, &minus_one);
+    MPNumber zero = mp_new();
+    MPNumber one = mp_new();
+    MPNumber minus_one = mp_new();
+    mp_set_from_integer(0, &zero);
+    mp_set_from_integer(1, &one);
+    mp_set_from_integer(-1, &minus_one);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

File: src/unittest.c [hunk 6]
```diff
@@ -766,6 +749,8 @@ static void test_mp()
 	try("mp_is_less_equal(-1, -1)", mp_is_less_equal (&minus_one, &minus_one), true);
 	try("mp_is_less_equal(-1, 0)", mp_is_less_equal (&minus_one, &zero), true);
 	try("mp_is_less_equal(-1, 1)", mp_is_less_equal (&minus_one, &one), true);
+
+    mp_clear(&zero); mp_clear(&one); mp_clear(&minus_one);
```
Comment (sc0w) [inline src/unittest.c]: wrong indentation

**Comment by sc0w** [comment]
please add packages in alphabetical order in https://github.com/mate-desktop/mate-calc/pull/138/commits/78653036265027251c47b734e11ef0c546bbed89

**Comment by mbkma** [comment]
> please add packages in alphabetical order in [7865303](https://github.com/mate-desktop/mate-calc/commit/78653036265027251c47b734e11ef0c546bbed89)

@sc0w done :)

**Comment by raveit65** [comment]
Yes.

**Pull request merged**